### PR TITLE
Add mackerel-plugin-solr

### PR DESCRIPTION
**Pull request by supercaracal**
A mackerel plugin for [Apache Solr](http://lucene.apache.org/solr/)

I'll implement some test code later. I'm sorry. @Songmu 

#### Output Sample

- host ... 192.168.33.10
- core ... core1

```
$ go run mackerel-plugin-solr/solr.go -host=192.168.33.10
solr.core1.segmentCount.core1_segmentCount    3       1448706470
solr.core1.sizeInBytes.core1_sizeInBytes      105116679       1448706470
solr.core1.timeouts.core1_timeouts_updatejson 0       1448706470
solr.core1.timeouts.core1_timeouts_select     0       1448706470
solr.core1.timeouts.core1_timeouts_updatejsondocs     0       1448706470
solr.core1.timeouts.core1_timeouts_get        0       1448706470
solr.core1.timeouts.core1_timeouts_updatecsv  0       1448706470
solr.core1.timeouts.core1_timeouts_replication        0       1448706470
solr.core1.timeouts.core1_timeouts_update     0       1448706470
solr.core1.timeouts.core1_timeouts_dataimport 0       1448706470
solr.core1.95thPcRequestTime.core1_95thPcRequestTime_updatejson       0       1448706470
solr.core1.95thPcRequestTime.core1_95thPcRequestTime_select   1.699699        1448706470
solr.core1.95thPcRequestTime.core1_95thPcRequestTime_updatejsondocs   0       1448706470
solr.core1.95thPcRequestTime.core1_95thPcRequestTime_get      0       1448706470
solr.core1.95thPcRequestTime.core1_95thPcRequestTime_updatecsv        0       1448706470
solr.core1.95thPcRequestTime.core1_95thPcRequestTime_replication      2.921822        1448706470
solr.core1.95thPcRequestTime.core1_95thPcRequestTime_update   0       1448706470
solr.core1.95thPcRequestTime.core1_95thPcRequestTime_dataimport       37.191631       1448706470
solr.core1.requests.core1_requests_updatejson 0       1448706470
solr.core1.requests.core1_requests_select     217     1448706470
solr.core1.requests.core1_requests_updatejsondocs     0       1448706470
solr.core1.requests.core1_requests_get        0       1448706470
solr.core1.requests.core1_requests_updatecsv  0       1448706470
solr.core1.requests.core1_requests_replication        16      1448706470
solr.core1.requests.core1_requests_update     0       1448706470
solr.core1.requests.core1_requests_dataimport 14      1448706470
solr.core1.5minRateReqsPerSecond.core1_5minRateReqsPerSecond_updatejson       0       1448706470
solr.core1.5minRateReqsPerSecond.core1_5minRateReqsPerSecond_select   0.016260        1448706470
solr.core1.5minRateReqsPerSecond.core1_5minRateReqsPerSecond_updatejsondocs   0       1448706470
solr.core1.5minRateReqsPerSecond.core1_5minRateReqsPerSecond_get      0       1448706470
solr.core1.5minRateReqsPerSecond.core1_5minRateReqsPerSecond_updatecsv        0       1448706470
solr.core1.5minRateReqsPerSecond.core1_5minRateReqsPerSecond_replication      0.000000        1448706470
solr.core1.5minRateReqsPerSecond.core1_5minRateReqsPerSecond_update   0       1448706470
solr.core1.5minRateReqsPerSecond.core1_5minRateReqsPerSecond_dataimport       0.000000        1448706470
solr.core1.15minRateReqsPerSecond.core1_15minRateReqsPerSecond_updatejson     0       1448706470
solr.core1.15minRateReqsPerSecond.core1_15minRateReqsPerSecond_select 0.017310        1448706470
solr.core1.15minRateReqsPerSecond.core1_15minRateReqsPerSecond_updatejsondocs 0       1448706470
solr.core1.15minRateReqsPerSecond.core1_15minRateReqsPerSecond_get    0       1448706470
solr.core1.15minRateReqsPerSecond.core1_15minRateReqsPerSecond_updatecsv      0       1448706470
solr.core1.15minRateReqsPerSecond.core1_15minRateReqsPerSecond_replication    0.000000        1448706470
solr.core1.15minRateReqsPerSecond.core1_15minRateReqsPerSecond_update 0       1448706470
solr.core1.15minRateReqsPerSecond.core1_15minRateReqsPerSecond_dataimport     0.000000        1448706470
solr.core1.medianRequestTime.core1_medianRequestTime_updatejson       0       1448706470
solr.core1.medianRequestTime.core1_medianRequestTime_select   0.853979        1448706470
solr.core1.medianRequestTime.core1_medianRequestTime_updatejsondocs   0       1448706470
solr.core1.medianRequestTime.core1_medianRequestTime_get      0       1448706470
solr.core1.medianRequestTime.core1_medianRequestTime_updatecsv        0       1448706470
solr.core1.medianRequestTime.core1_medianRequestTime_replication      0.850494        1448706470
solr.core1.medianRequestTime.core1_medianRequestTime_update   0       1448706470
solr.core1.medianRequestTime.core1_medianRequestTime_dataimport       0.132340        1448706470
solr.core1.hits.core1_hits_filterCache        32      1448706470
solr.core1.hits.core1_hits_perSegFilter       0       1448706470
solr.core1.hits.core1_hits_queryResultCache   190     1448706470
solr.core1.hits.core1_hits_documentCache      487     1448706470
solr.core1.hits.core1_hits_fieldValueCache    0       1448706470
solr.core1.size.core1_size_filterCache        14      1448706470
solr.core1.size.core1_size_perSegFilter       0       1448706470
solr.core1.size.core1_size_queryResultCache   26      1448706470
solr.core1.size.core1_size_documentCache      138     1448706470
solr.core1.size.core1_size_fieldValueCache    0       1448706470
solr.core1.evictions.core1_evictions_filterCache      0       1448706470
solr.core1.evictions.core1_evictions_perSegFilter     0       1448706470
solr.core1.evictions.core1_evictions_queryResultCache 0       1448706470
solr.core1.evictions.core1_evictions_documentCache    0       1448706470
solr.core1.evictions.core1_evictions_fieldValueCache  0       1448706470
solr.core1.warmupTime.core1_warmupTime_filterCache    0       1448706470
solr.core1.warmupTime.core1_warmupTime_perSegFilter   0       1448706470
solr.core1.warmupTime.core1_warmupTime_queryResultCache       0       1448706470
solr.core1.warmupTime.core1_warmupTime_documentCache  0       1448706470
solr.core1.warmupTime.core1_warmupTime_fieldValueCache        0       1448706470
solr.core1.docsCount.core1_numDocs    120550  1448706470
solr.core1.docsCount.core1_deletedDocs        0       1448706470
solr.core1.indexHeapUsageBytes.core1_indexHeapUsageBytes      -1      1448706470
solr.core1.errors.core1_errors_updatejson     0       1448706470
solr.core1.errors.core1_errors_select 1       1448706470
solr.core1.errors.core1_errors_updatejsondocs 0       1448706470
solr.core1.errors.core1_errors_get    0       1448706470
solr.core1.errors.core1_errors_updatecsv      0       1448706470
solr.core1.errors.core1_errors_replication    0       1448706470
solr.core1.errors.core1_errors_update 0       1448706470
solr.core1.errors.core1_errors_dataimport     0       1448706470
solr.core1.75thPcRequestTime.core1_75thPcRequestTime_updatejson       0       1448706470
solr.core1.75thPcRequestTime.core1_75thPcRequestTime_select   1.164905        1448706470
solr.core1.75thPcRequestTime.core1_75thPcRequestTime_updatejsondocs   0       1448706470
solr.core1.75thPcRequestTime.core1_75thPcRequestTime_get      0       1448706470
solr.core1.75thPcRequestTime.core1_75thPcRequestTime_updatecsv        0       1448706470
solr.core1.75thPcRequestTime.core1_75thPcRequestTime_replication      2.010164        1448706470
solr.core1.75thPcRequestTime.core1_75thPcRequestTime_update   0       1448706470
solr.core1.75thPcRequestTime.core1_75thPcRequestTime_dataimport       1.188915        1448706470
solr.core1.99thPcRequestTime.core1_99thPcRequestTime_updatejson       0       1448706470
solr.core1.99thPcRequestTime.core1_99thPcRequestTime_select   6.464342        1448706470
solr.core1.99thPcRequestTime.core1_99thPcRequestTime_updatejsondocs   0       1448706470
solr.core1.99thPcRequestTime.core1_99thPcRequestTime_get      0       1448706470
solr.core1.99thPcRequestTime.core1_99thPcRequestTime_updatecsv        0       1448706470
solr.core1.99thPcRequestTime.core1_99thPcRequestTime_replication      2.921822        1448706470
solr.core1.99thPcRequestTime.core1_99thPcRequestTime_update   0       1448706470
solr.core1.99thPcRequestTime.core1_99thPcRequestTime_dataimport       37.191631       1448706470
solr.core1.lookups.core1_lookups_filterCache  39      1448706470
solr.core1.lookups.core1_lookups_perSegFilter 0       1448706470
solr.core1.lookups.core1_lookups_queryResultCache     216     1448706470
solr.core1.lookups.core1_lookups_documentCache        625     1448706470
solr.core1.lookups.core1_lookups_fieldValueCache      0       1448706470
solr.core1.avgRequestsPerSecond.core1_avgRequestsPerSecond_updatejson 0       1448706470
solr.core1.avgRequestsPerSecond.core1_avgRequestsPerSecond_select     0.000163        1448706470
solr.core1.avgRequestsPerSecond.core1_avgRequestsPerSecond_updatejsondocs     0       1448706470
solr.core1.avgRequestsPerSecond.core1_avgRequestsPerSecond_get        0       1448706470
solr.core1.avgRequestsPerSecond.core1_avgRequestsPerSecond_updatecsv  0       1448706470
solr.core1.avgRequestsPerSecond.core1_avgRequestsPerSecond_replication        0.000012        1448706470
solr.core1.avgRequestsPerSecond.core1_avgRequestsPerSecond_update     0       1448706470
solr.core1.avgRequestsPerSecond.core1_avgRequestsPerSecond_dataimport 0.000011        1448706470
solr.core1.avgTimePerRequest.core1_avgTimePerRequest_updatejson       0       1448706470
solr.core1.avgTimePerRequest.core1_avgTimePerRequest_select   9.533985        1448706470
solr.core1.avgTimePerRequest.core1_avgTimePerRequest_updatejsondocs   0       1448706470
solr.core1.avgTimePerRequest.core1_avgTimePerRequest_get      0       1448706470
solr.core1.avgTimePerRequest.core1_avgTimePerRequest_updatecsv        0       1448706470
solr.core1.avgTimePerRequest.core1_avgTimePerRequest_replication      1.834254        1448706470
solr.core1.avgTimePerRequest.core1_avgTimePerRequest_update   0       1448706470
solr.core1.avgTimePerRequest.core1_avgTimePerRequest_dataimport       3.325568        1448706470
solr.core1.999thPcRequestTime.core1_999thPcRequestTime_updatejson     0       1448706470
solr.core1.999thPcRequestTime.core1_999thPcRequestTime_select 8.706049        1448706470
solr.core1.999thPcRequestTime.core1_999thPcRequestTime_updatejsondocs 0       1448706470
solr.core1.999thPcRequestTime.core1_999thPcRequestTime_get    0       1448706470
solr.core1.999thPcRequestTime.core1_999thPcRequestTime_updatecsv      0       1448706470
solr.core1.999thPcRequestTime.core1_999thPcRequestTime_replication    2.921822        1448706470
solr.core1.999thPcRequestTime.core1_999thPcRequestTime_update 0       1448706470
solr.core1.999thPcRequestTime.core1_999thPcRequestTime_dataimport     37.191631       1448706470
solr.core1.hitratio.core1_hitratio_filterCache        0.820000        1448706470
solr.core1.hitratio.core1_hitratio_perSegFilter       0       1448706470
solr.core1.hitratio.core1_hitratio_queryResultCache   0.880000        1448706470
solr.core1.hitratio.core1_hitratio_documentCache      0.780000        1448706470
solr.core1.hitratio.core1_hitratio_fieldValueCache    0       1448706470
solr.core1.inserts.core1_inserts_filterCache  14      1448706470
solr.core1.inserts.core1_inserts_perSegFilter 0       1448706470
solr.core1.inserts.core1_inserts_queryResultCache     26      1448706470
solr.core1.inserts.core1_inserts_documentCache        138     1448706470
solr.core1.inserts.core1_inserts_fieldValueCache      0       1448706470
```

### CODE REVIEW

File: mackerel-plugin-solr/solr.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+			graphdef[fmt.Sprintf("%s.%s.%s", s.Prefix, core, key)] = mp.Graphs{
+				Label: fmt.Sprintf("%s %s", labelCore, metricLabel),
+				Unit:  "integer",
+				Metrics: [](mp.Metrics){
+					mp.Metrics{Name: core + "_" + key, Label: metricLabel, Diff: false},
+				},
+			}
+		}
+
+		for _, key := range queryHandlerStatKeys {
```
Comment (supercaracal) [inline mackerel-plugin-solr/solr.go]: This graph definition have not any "Label" parameters effect in roops. The cause remains unknown. Hmm...

File: mackerel-plugin-solr/solr.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+				)
+			}
+			graphdef[fmt.Sprintf("%s.%s.%s", s.Prefix, core, key)] = mp.Graphs{
+				Label:   fmt.Sprintf("%s %s", labelCore, strings.Title(key)),
+				Unit:    "integer",
+				Metrics: metrics,
+			}
+		}
+
+		for _, key := range cacheStatKeys {
```
Comment (supercaracal) [inline mackerel-plugin-solr/solr.go]: Same above.

File: mackerel-plugin-solr/solr.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+		graphdef[fmt.Sprintf("%s.%s.docsCount", s.Prefix, core)] = mp.Graphs{
+			Label: fmt.Sprintf("%s DocsCount", labelCore),
+			Unit:  "integer",
+			Metrics: [](mp.Metrics){
+				mp.Metrics{Name: core + "_numDocs", Label: "NumDocs", Diff: false},
+				mp.Metrics{Name: core + "_deletedDocs", Label: "DeletedDocs", Diff: false},
+			},
+		}
+
+		for _, key := range []string{"indexHeapUsageBytes", "segmentCount", "sizeInBytes"} {
```
Comment (supercaracal) [inline mackerel-plugin-solr/solr.go]: This label definition is worked.

File: mackerel-plugin-solr/solr.go [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+// GraphDefinition interface for mackerelplugin
+func (s SolrPlugin) GraphDefinition() map[string](mp.Graphs) {
+	graphdef := make(map[string](mp.Graphs))
+	var labelCore string
+
+	for _, core := range s.Cores {
+		labelCore = strings.Title(core)
+
+		graphdef[fmt.Sprintf("%s.%s.docsCount", s.Prefix, core)] = mp.Graphs{
```
Comment (supercaracal) [inline mackerel-plugin-solr/solr.go]: This label definition is worked.

**Comment by supercaracal** [comment]
(私は英語ができないので日本語を添えさせてください m(_ _)m)

**Comment by Songmu** [comment]
@supercaracal Do you plan to implement some test code on this pull request or not?

Can I start to review this pull request?

**Comment by supercaracal** [comment]
@Songmu Yes. Please start to review.

I will try implement test codes at other pull request. I'm sorry.

**Comment by Songmu** [comment]
I do merge it, but we do not include it into core package (yum and deb) until test codes are commited. We wait your following pull request.

Thank you!

**Comment by supercaracal** [comment]
@Songmu It’s noted with thanks. :+1:

**Pull request merged**